### PR TITLE
chore(buffers): Rework gauge metric handling

### PR DESCRIPTION
**Pull request by bruceg**
## Summary

This is a step-by-step refactoring of the changes made in #23453 and #23507 in order to remove the need for the shared `BUFFER_COUNTERS` dashmap. While the whole change is rather large, I attempted to keep each commit from altering externally-visible behavior, with the exception of additional labels in the buffer create gauges. As such, reviewing each individual commit is likely more useful than reviewing the whole code change.

If you compare this PR to the code base [starting immediately before the two PRs referenced above](https://github.com/vectordotdev/vector/compare/c7eaf6529079fb1070f913aaf2a3d5145c5f4001...bruceg/rework-buffer-counters) (have to manually scroll down to the buffer source files, the direct GitHub link doesn't quite work), the improvements are evident:

1. The gauges now have an additional `buffer_id` label to disambiguate the metrics with the same ID but different stages. Note that the buffers are created in the context of a component and, as such, should (may?) already have the component ID in their tags. If not, this is a real bug, maybe even the real bug. Since the `buffer_id` corresponds to a component ID, maybe that particular tag name is less than ideal.
2. The buffer usage tracker also maintains a `current` values set, allowing us to `set` the gauges instead of `incrementing` them, resulting in more reliable values there.
3. All the counters are updated using a `saturating_add`, which prevents them from overflowing, and a `clamp`, which prevents them from going negative.

Note: This is a rework of #23542 to fix some bugs I introduced in my line of changes, so see that PR for more context.

## Vector configuration

N/A

## How did you test this PR?

Tested using the unit tests, all of the above should be behavior-preserving.

## Change Type
- [ ] Bug fix
- [ ] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?
- [ ] Yes
- [ ] No

## Does this PR include user facing changes?

- [ ] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the `no-changelog` label to this PR.

## Notes
- Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).
- Do not hesitate to use `@vectordotdev/vector` to reach out to us regarding this PR.
- Some CI checks run only after we manually approve them.
  - We recommend adding a `pre-push` hook, please see [this template](https://github.com/vectordotdev/vector/blob/master/CONTRIBUTING.md#Pre-push).
  - Alternatively, we recommend running the following locally before pushing to the remote branch:
    - `cargo fmt --all`
    - `cargo clippy --workspace --all-targets -- -D warnings`
    - `cargo nextest run --workspace` (alternatively, you can run `cargo test --all`)
- After a review is requested, please avoid force pushes to help us review incrementally.
  - Feel free to push as many commits as you want. They will be squashed into one before merging.
  - For example, you can run `git merge origin master` and `git push`.
- If this PR introduces changes Vector dependencies (modifies `Cargo.lock`), please
  run `cargo vdev build licenses` to regenerate the [license inventory](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv) and commit the changes (if any). More details [here](https://crates.io/crates/dd-rust-license-tool).

### CODE REVIEW

File: lib/vector-buffers/src/buffer_usage_data.rs [hunk 1]
```diff
@@ -19,18 +19,21 @@ use crate::{
 static BUFFER_COUNTERS: LazyLock<DashMap<(String, usize), (AtomicU64, AtomicU64)>> =
     LazyLock::new(DashMap::new);
 
+fn u64_to_i64(value: u64) -> i64 {
```
Comment (graphcareful) [inline lib/vector-buffers/src/buffer_usage_data.rs]: Maybe this function belongs in a util module or something of the sort? If one already exists

File: lib/vector-buffers/src/cast_utils.rs [hunk 2]
```diff
@@ -1,10 +0,0 @@
-// Maximum integer value that can be represented exactly in f64 (2^53).
```
Comment (graphcareful) [inline lib/vector-buffers/src/cast_utils.rs]: Nice change, makes sense to me

**Review by graphcareful — APPROVED** [review]
Just want to leave a note, this PR was super easy to review when looking at it commit at a time.

File: lib/vector-buffers/src/buffer_usage_data.rs [hunk 3]
```diff
@@ -118,13 +142,15 @@ impl BufferUsageHandle {
     /// This represents the events being sent into the buffer.
     pub fn increment_received_event_count_and_byte_size(&self, count: u64, byte_size: u64) {
         self.state.received.increment(count, byte_size);
+        self.state.current.increment(count, byte_size);
     }
 
     /// Increments the number of events (and their total size) sent by this buffer component.
     ///
     /// This represents the events being read out of the buffer.
     pub fn increment_sent_event_count_and_byte_size(&self, count: u64, byte_size: u64) {
```
Comment (vparfonov) [inline lib/vector-buffers/src/buffer_usage_data.rs:165]: nit:
```
if count == 0 && byte_size == 0 {
    return; 
}
```

File: lib/vector-buffers/src/buffer_usage_data.rs [hunk 4]
```diff
@@ -313,3 +351,36 @@ impl BufferUsage {
...
+            });
+            handles.push(handle);
+        }
+
+        for handle in handles {
+            handle.join().unwrap();
+        }
+
+        assert_eq!(counter.load(Ordering::Relaxed), 0);
+    }
```
Comment (vparfonov) [inline lib/vector-buffers/src/buffer_usage_data.rs:437]: suggestion add test to check negative value: 
```
#[test]
fn test_update_counter_prevents_negatives() {
  let counter = AtomicU64::new(100);
  
  update_counter(&counter, -50);
  assert_eq!(counter.load(Ordering::Relaxed), 50);
        
  update_counter(&counter, -100);
  assert_eq!(counter.load(Ordering::Relaxed), 0);

  update_counter(&counter, -50);
  assert_eq!(counter.load(Ordering::Relaxed), 0);
}
```

**Comment by vparfonov** [comment]
LGTM, I closed my PR, this solution looks better.

**Comment by pront** [comment]
> LGTM, I closed my PR, this solution looks better.

@vparfonov perhaps you can help us verify with your setup that this PR yields the same metrics as `origin/master`? Thanks!

File: lib/vector-buffers/src/buffer_usage_data.rs (hunk 1 above)
Comment (bruceg) [reply lib/vector-buffers/src/buffer_usage_data.rs]: Maybe once it's used in more than one place, yes.

File: lib/vector-buffers/src/buffer_usage_data.rs (hunk 4 above)
Comment (bruceg) [reply lib/vector-buffers/src/buffer_usage_data.rs:437]: Added this and an overflow test in b840c9e2a

File: lib/vector-buffers/src/buffer_usage_data.rs (hunk 3 above)
Comment (bruceg) [reply lib/vector-buffers/src/buffer_usage_data.rs:165]: Makes sense, done in 8a26a4095

File: lib/vector-buffers/src/buffer_usage_data.rs [hunk 5]
```diff
@@ -15,6 +15,25 @@ use crate::{
     spawn_named,
 };
 
+fn u64_to_i64(value: u64) -> i64 {
+    // If we ever have to deal with more than 9,223,372,036,854,775,808 events or bytes _in a single
+    // update_, this conversion will break counters. That number is effectively impossibly large,
```
Comment (pront) [inline lib/vector-buffers/src/buffer_usage_data.rs]: This comment basically describes what `try_from` does below. Alternatively, let's remove it and handle the Result below. By emitting an `error!` (instead of silently defaulting to `i64::MAX`) we might catch some rare bug.

File: lib/vector-buffers/src/buffer_usage_data.rs (hunk 1 above)
Comment (pront) [reply lib/vector-buffers/src/buffer_usage_data.rs]: We had `cast_utils.rs` for this but this since it's not used in multiple places, it is simpler to put it here.

File: lib/vector-buffers/src/internal_events.rs [hunk 6]
```diff
@@ -156,11 +157,24 @@ impl InternalEvent for BufferEventsDropped {
             "intentional" => intentional_str,
         )
         .increment(self.count);
-
-        let count_delta = i64::try_from(self.count).unwrap_or(i64::MAX);
-        let bytes_delta = i64::try_from(self.byte_size).unwrap_or(i64::MAX);
-
-        update_buffer_gauge(&self.buffer_id, self.idx, -count_delta, -bytes_delta);
+        counter!(
```
[the remark this answers is not in the document]
Comment (pront) [reply lib/vector-buffers/src/internal_events.rs:161]: Is it really?

[the remark this answers is not in the document]
Comment (bruceg) [reply lib/vector-buffers/src/internal_events.rs:161]: Ah, yes it is, I'll fix that.

[the remark this answers is not in the document]
Comment (bruceg) [reply lib/vector-buffers/src/internal_events.rs:161]: Ah, no, it is no longer missing, that deficiency has been fixed after this review.

File: lib/vector-buffers/src/buffer_usage_data.rs (hunk 5 above)
Comment (bruceg) [reply lib/vector-buffers/src/buffer_usage_data.rs]: Note that this is also called in `CategoryMetrics::{in,de}crement` as well. This will need different handling.

By that token, however, we should also not `clamp`, but that was kinda the point of the original work.

Not sure how to proceed here.

Comment (pront) [reply lib/vector-buffers/src/buffer_usage_data.rs]: I was thinking of the following:

```rust
let result = match i64::try_from(value) {
    Ok(v) => v,
    Err(e) => {
        error!("Defaulting to i64::MAX. Failed to convert value to i64: {e}");
        i64::MAX
    }
};
```

Comment (bruceg) [reply lib/vector-buffers/src/buffer_usage_data.rs]: Would you like to see the same pattern replace the `clamp` as well?

Comment (bruceg) [reply lib/vector-buffers/src/buffer_usage_data.rs]: Reworked in c5ecfcb1d

Comment (pront) [reply lib/vector-buffers/src/buffer_usage_data.rs]: Nice, I like this version better.

**Review by pront — APPROVED** [review]
LGTM. It would also be nice to get validation per https://github.com/vectordotdev/vector/pull/23561#issuecomment-3184309835

**Comment by vparfonov** [comment]
> LGTM. It would also be nice to get validation per [#23561 (comment)](https://github.com/vectordotdev/vector/pull/23561#issuecomment-3184309835)

Sure, I will test tomorrow morning and back here with results

**Comment by vparfonov** [comment]
@bruceg @pront I've tested this changes, works just fine. Thanks

**Comment by bruceg** [comment]
@vparfonov this testing you are able to do, would there be any way that could be worked into a unit test? It would be good to add that as evidence things are working as intended.

**Comment by pront** [comment]
> @bruceg @pront I've tested this changes, works just fine. Thanks

👍 thanks

> @vparfonov this testing you are able to do, would there be any way that could be worked into a unit test? It would be good to add that as evidence things are working as intended.

we can do it as a follow-up if possible.
